File: benchmark/HashTableBenchmark.java
import java.util.HashMap;
import java.util.Map;

public class HashTableBenchmark {
    public static void main(String[] args) {
        Map<Long, Long> map = new HashMap<>();
        long startTime = System.nanoTime();
        for (int i = 0; i < 1000000; i++) {
            long key = ((long) i << 32) | i;
            map.put(key, (long)i);
        }
        for (int i = 0; i < 1000000; i++) {
            long key = ((long) i << 32) | i;
            long value = map.get(key);
            map.remove(key);
            long newKey = ((long) i << 32) | (i + 1);
            map.put(newKey, value);
        }
        long endTime = System.nanoTime();
        long duration = (endTime - startTime);
        System.out.println("Time taken to insert, retrieve, delete, and add 1 million entries: " + duration + " nanoseconds");
    }
}

File: benchmark/HashTableBenchmark.py
import time

def hash(x, z):
    return (x << 32) | z

table = {}
start_time = time.time_ns()
for i in range(1000000):
    key = hash(i, i)
    table[key] = i
for i in range(1000000):
    key = hash(i, i)
    value = table[key]
    del table[key]
    new_key = hash(i, i + 1)
    table[new_key] = value
end_time = time.time_ns()
duration = end_time - start_time
print(f"Time taken to insert, retrieve, delete, and add 1 million entries: {duration} nanoseconds")
